File: src/notification_tray/components/notifier.py
import logging
from datetime import UTC, datetime
from functools import partial
from pathlib import Path

from PyQt5.QtCore import QObject, pyqtSignal
from PyQt5.QtGui import QScreen
from PyQt5.QtMultimedia import QSound
from PyQt5.QtWidgets import QApplication

from notification_tray.components.notification_service import \
    NotificationCloseReason
from notification_tray.components.notification_widget import NotificationWidget
from notification_tray_stubs.notification import (CachedNotification,
                                                  NotificationFolder)
from notification_tray.utils import settings
from notification_tray.utils.settings import (get_do_not_disturb,
                                              get_notification_backoff_minutes,
                                              is_do_not_disturb_active)

logger = logging.getLogger(__name__)


class Notifier(QObject):
    action_invoked = pyqtSignal(int, str)
    notification_displayed = pyqtSignal(int, str, str, str)
    notification_closed = pyqtSignal(int, int, str, bool)
    last_notified: dict[Path, int] = {}
    started_at = datetime.now(UTC)
    notification_widgets = dict[int, NotificationWidget]()
    offset = 0

    def __init__(
        self,
        root_path: Path,
        do_not_disturb: settings.Cache,
        notification_backoff_minutes: dict[Path, int],
        notification_cache: NotificationFolder,
        parent: QObject | None = None,
    ) -> None:
        self.started_at = datetime.now(UTC)
        self.root_path = root_path
        self.do_not_disturb = do_not_disturb
        self.notification_backoff_minutes = notification_backoff_minutes
        self.notification_cache = notification_cache
        logger.info(f"Started notifier with root path {root_path}")
        super().__init__(parent)

    def show_or_queue_notification(self, notification_widget: NotificationWidget):
        logger.info(
            f"Got request to show notification {notification_widget.data['id']}"
        )
        self.notification_widgets[notification_widget.data["id"]] = notification_widget
        screen = QApplication.primaryScreen()
        match screen:
            case None:
                raise RuntimeError("No screen")
            case QScreen():
                screen = screen.availableGeometry()

        if (x := notification_widget.data["hints"].get("x")) is not None and (
            y := notification_widget.data["hints"].get("y")
        ) is not None:
            logger.debug(
                f"Notification {notification_widget.data['id']} has x,y hints ({x},{y}). Moving there"
            )
            notification_widget.move(x, y)

        else:
            if screen.height() - notification_widget.height() - self.offset > 0:
                logger.debug(
                    f"Moving notification {notification_widget.data['id']} to next available place"
                )
                notification_widget.move(
                    screen.width() - notification_widget.width(),
                    screen.height() - notification_widget.height() - self.offset,
                )
                self.offset += notification_widget.height() + 10
                notification_widget.show()
                notification_widget.displayed.connect(
                    lambda: self.notification_displayed.emit(
                        max(0, notification_widget.data["id"]),
                        notification_widget.data["app_name"],
                        notification_widget.data["summary"],
                        notification_widget.data["body"],
                    )
                )
                notification_widget.displayed.emit()
            else:
                logger.debug(
                    f"No screen space left for notification {notification_widget.data['id']}. Queuing"
                )

    def close_notification(
        self,
        notification_id_or_widget: int | NotificationWidget,
        reason: int,
        is_batch: bool = False,
    ):
        reason = NotificationCloseReason(reason)
        match notification_id_or_widget:
            case NotificationWidget():
                notification_widget = notification_id_or_widget
            case int():
                notification_widget = self.notification_widgets.get(
                    notification_id_or_widget
                )
                if notification_widget is None:
                    logger.error(
                        f"Could not find notification with id {notification_id_or_widget}. Skipping"
                    )
                    return

        logger.info(
            f"Closing notification {notification_widget.data['id']}. Reason {reason.name}"
        )
        if notification_widget.isVisible():
            notification_widget.close()
        if (
            reason != NotificationCloseReason.CLOSED_BY_CALL_TO_CLOSENOTIFICATION
        ):  # signal already emitted by NotificationService, no need to re-emit
            self.notification_closed.emit(
                notification_widget.data["id"],
                reason.value,
                str(notification_widget.data["path"]),
                notification_widget.data["at"] >= self.started_at
                and not is_batch
                and notification_widget.data["id"] >= 0,
            )
        self.offset = sum(
            widget.height() + 10
            for widget in self.notification_widgets.values()
            if widget.isVisible()
        )
        logger.info("Displaying queued notifications")
        for notification_widget in self.notification_widgets.values():
            if not notification_widget.was_displayed:
                self.show_or_queue_notification(notification_widget)

    def notify(
        self,
        notification: CachedNotification,
        *notifications: CachedNotification,
        is_batch: bool = False,
    ) -> None:
        logger.info(
            f"Got request to display {len([notification, *notifications])} notifications"
        )
        logger.info(
            "Checking which notifications should be displayed based on urgency, DnD and backoff settings"
        )
        all_notifications = [
            n
            for n in [notification, *notifications]
            if not n.get("trashed")
            and (
                n["hints"].get("urgency") == 2
                or (
                    not is_do_not_disturb_active(
                        self.root_path, n["path"].parent, self.do_not_disturb
                    )
                    and (
                        is_batch
                        or get_notification_backoff_minutes(
                            self.root_path,
                            n["path"].parent,
                            self.notification_backoff_minutes,
                        )
                        == 0
                    )
                )
            )
        ]
        logger.info(f"{len(all_notifications)} to display")
        if all_notifications:
            try:
                app_name = all_notifications[-1]["app_name"]
                content = (
                    all_notifications[-1]["body"]
                    if len(all_notifications) == 1
                    else "\n---\n".join(
                        map(
                            lambda d: "\n".join(
                                filter(
                                    None,
                                    [
                                        d["summary"],
                                        d["body"],
                                    ],
                                )
                            ),
                            all_notifications,
                        ),
                    )
                )
                # Truncate content if it's too long
                if len(content) >= 1000:
                    content = content[:997] + "..."

                self.play_notification_sound(all_notifications[-1])
                notification_widget = NotificationWidget(
                    CachedNotification(
                        notification,
                        summary=(
                            all_notifications[-1]["summary"]
                            if len(all_notifications) == 1
                            else f"{len(all_notifications)} new notifications from {app_name}"
                        ),
                        body=content,
                        expire_timeout=(
                            0
                            if any(
                                notification["expire_timeout"] == 0
                                for notification in all_notifications
                            )
                            else max(
                                -1,
                                sum(
                                    notification["expire_timeout"]
                                    for notification in all_notifications
                                ),
                            )
                        ),
                    )
                )
                notification_widget.closed.connect(
                    partial(
                        self.close_notification, notification_widget, is_batch=is_batch
                    )
                )
                notification_widget.action_invoked.connect(
                    lambda key: (
                        self.action_invoked.emit(notification_widget.data["id"], key)
                        if notification_widget.data["at"] >= self.started_at
                        else None
                    )
                )
                self.show_or_queue_notification(notification_widget)
                for notification in all_notifications:
                    parent = notification["path"].parent
                    logger.debug(f"Updating last_notified timestamp for {parent}")
                    self.last_notified[parent] = max(
                        self.last_notified.setdefault(parent, -1), notification["id"]
                    )

            except Exception as e:
                logger.exception(f"Unable to read notifications: {e!s}")
                self.notify(
                    CachedNotification(
                        summary="Error",
                        body=f"Unable to read notifications: {e!s}",
                        app_icon="error",
                        at=datetime.now(UTC),
                        hints={"sound-name": "dialog-error"},
                        actions={},
                        app_name="notification-tray",
                        expire_timeout=-1,
                        id=-1,
                        replaces_id=0,
                        path=self.root_path / "error.json",
                    )
                )

    def batch_notify(self):
        def process(folder: NotificationFolder):
            logger.debug(f"Processing batch notifications for folder {folder['path']}")
            self.last_notified.setdefault(folder["path"], -1)
            new_notifications = list[CachedNotification]()
            logger.debug(f"{len(folder['notifications'])} notifications to process")
            for notification in folder["notifications"].values():
                at = notification["at"]
                id = notification["id"]
                logger.debug(f"Processing notification {id}")
                notification_backoff_minutes = get_notification_backoff_minutes(
                    self.root_path, folder["path"], self.notification_backoff_minutes
                )
                minutes_since_last_notification = (
                    datetime.now(UTC) - at
                ).total_seconds() // 60
                if not is_do_not_disturb_active(
                    self.root_path, folder["path"], self.do_not_disturb
                ) and (
                    (
                        notification_backoff_minutes > 0
                        and minutes_since_last_notification
                        <= notification_backoff_minutes
                    )
                    or (
                        # we just came back from DnD
                        (
                            do_not_disturb := get_do_not_disturb(
                                self.root_path, folder["path"], self.do_not_disturb
                            )
                        )
                        and do_not_disturb >= self.started_at
                        and at >= do_not_disturb
                        and id > self.last_notified[folder["path"]]
                    )
                ):
                    logger.debug("Adding notification to batch")
                    new_notifications.append(notification)
                else:
                    logger.debug("Skipping notification due to DnD or backoff settings")

            if new_notifications:
                self.notify(*new_notifications, is_batch=True)
            for folder in folder["folders"].values():
                process(folder)

        process(self.notification_cache)

    def play_notification_sound(self, notification: CachedNotification):
        if notification.get("suppress-sound"):
            logger.debug(
                f"Notification {notification['id']} has suppress-sound hint. Not playing sound"
            )
        else:
            audio_path: str | None = None
            if sound_file := notification["hints"].get("sound-file"):
                audio_path = sound_file
                logger.debug(
                    f"Getting sound for {notification['id']} from sound-file hint"
                )
            elif sound_name := notification["hints"].get("sound-name"):
                audio_path = f"/usr/share/sounds/freedesktop/{sound_name}.oga"
                logger.debug(
                    f"Getting sound for {notification['id']} from sound-name hint"
                )
            else:
                for folder in [
                    notification["path"].parent,
                    *map(
                        self.root_path.joinpath,
                        notification["path"].parent.relative_to(self.root_path).parents,
                    ),
                ]:
                    logger.debug(f"Checking {folder}")
                    if (notification_sound := (folder / ".notification.wav")).exists():
                        audio_path = str(notification_sound)
                        logger.debug(
                            f"Getting sound for {notification['id']} from .notification.wav file"
                        )
                        break
            if audio_path:
                logger.debug(f"Notification sound: {audio_path}")
                QSound.play(audio_path)
            else:
                logger.debug(
                    f"No notification sound found for notification {notification['id']}"
                )

File: src/notifier.cpp
#include "notifier.h"

#include "utils/logging.h"
#include "utils/settings.h"

#include <QApplication>
#include <QFile>
#include <QGuiApplication>
#include <QScreen>
#include <QSoundEffect>
#include <QTimer>
#include <QUrl>

#include <stdexcept>

static Logger logger = Logger::getLogger("Notifier");

Notifier::Notifier(const fs::path& root_path, const Cache& do_not_disturb,
                   const std::map<fs::path, int>& notification_backoff_minutes,
                   NotificationFolder& notification_cache, const QString& run_id, QObject* parent)
    : QObject(parent)
    , m_root_path(root_path)
    , m_do_not_disturb(do_not_disturb)
    , m_notification_backoff_minutes(notification_backoff_minutes)
    , m_notification_cache(notification_cache)
    , m_run_id(run_id) {
    m_started_at = QDateTime::currentDateTimeUtc();
    logger.info(QString("Started notifier with root path %1")
                    .arg(QString::fromStdString(root_path.string())));
}

void Notifier::notify(const CachedNotification& notification, bool is_batch) {
    std::vector<CachedNotification> vec = {notification};
    notify(vec, is_batch);
}

void Notifier::notify(const std::vector<CachedNotification>& notifications, bool is_batch) {
    try {
        logger.info(QString("Got request to display %1 notifications").arg(notifications.size()));

        std::vector<CachedNotification> to_display;
        for (const auto& n : notifications) {
            if (n.trashed)
                continue;

            int urgency = n.hints.value("urgency", 1).toInt();
            if (urgency == 2) {
                to_display.push_back(n);
                continue;
            }

            bool dnd_active =
                Settings::isDoNotDisturbActive(m_root_path, n.path.parent_path(), m_do_not_disturb);
            if (dnd_active && !is_batch)
                continue;

            int backoff = Settings::getNotificationBackoffMinutes(m_root_path, n.path.parent_path(),
                                                                  m_notification_backoff_minutes);
            if (!is_batch && backoff > 0)
                continue;

            to_display.push_back(n);
        }

        logger.info(QString("%1 to display").arg(to_display.size()));

        if (to_display.empty())
            return;

        const CachedNotification& last = to_display.back();

        QString summary =
            to_display.size() == 1
                ? last.summary
                : QString("%1 new notifications from %2").arg(to_display.size()).arg(last.app_name);

        QString body;
        if (to_display.size() == 1) {
            body = last.body;
        } else {
            QStringList parts;
            for (const auto& n : to_display) {
                QStringList content;
                if (!n.summary.isEmpty())
                    content << n.summary;
                if (!n.body.isEmpty())
                    content << n.body;
                parts << content.join("\n");
            }
            body = parts.join("\n---\n");
        }

        if (body.length() >= 1000) {
            body = body.left(997) + "...";
        }

        playNotificationSound(last);

        CachedNotification display_notif = last;
        display_notif.summary = summary;
        display_notif.body = body;

        NotificationWidget* widget = new NotificationWidget(display_notif);

        connect(widget, &NotificationWidget::closed, [this, widget, is_batch](int reason) {
            closeNotification(widget, static_cast<NotificationCloseReason>(reason), is_batch);
        });

        connect(widget, &NotificationWidget::snoozed,
                [this, widget](int duration_ms) { snoozeNotification(widget, duration_ms); });

        if (display_notif.notification_tray_run_id == m_run_id) {
            connect(widget, &NotificationWidget::actionInvoked, [this, widget](const QString& key) {
                emit actionInvoked(widget->data.id, key);
            });
        }

        showOrQueueNotification(widget);

        for (const auto& n : to_display) {
            m_last_notified[n.path.parent_path()] =
                std::max(m_last_notified[n.path.parent_path()], n.id);
        }
    } catch (const std::exception& e) {
        // Display error notification like Python does
        logger.error(QString("Unable to read notifications: %1").arg(e.what()));

        CachedNotification error_notif;
        error_notif.summary = "Error";
        error_notif.body = QString("Unable to read notifications: %1").arg(e.what());
        error_notif.app_icon = "error";
        error_notif.app_name = "notification-tray";
        error_notif.expire_timeout = -1;
        error_notif.id = -1;
        error_notif.replaces_id = 0;
        error_notif.at = QDateTime::currentDateTimeUtc();
        error_notif.hints["sound-name"] = "dialog-error";
        error_notif.notification_tray_run_id = m_run_id;
        error_notif.path = m_root_path / "error.json";

        // Recursively call notify with error notification
        notify(error_notif, false);
    }
}

void Notifier::showOrQueueNotification(NotificationWidget* widget) {
    logger.info(QString("Got request to show notification %1").arg(widget->data.id));

    m_notification_widgets[{widget->data.notification_tray_run_id, widget->data.id}] = widget;

    QScreen* screen = QGuiApplication::primaryScreen();
    if (!screen) {
        logger.error("No screen available");
        return;
    }

    QRect geometry = screen->availableGeometry();

    QVariant x_hint = widget->data.hints.value("x");
    QVariant y_hint = widget->data.hints.value("y");
    if (!x_hint.isNull() && !y_hint.isNull()) {
        widget->move(x_hint.toInt(), y_hint.toInt());
    } else {
        if (geometry.height() - widget->height() - m_offset > 0) {
            widget->move(geometry.width() - widget->width(),
                         geometry.height() - widget->height() - m_offset);
            m_offset += widget->height() + 10;
            widget->show();

            if (widget->data.notification_tray_run_id == m_run_id) {
                connect(widget, &NotificationWidget::displayed, [this, widget]() {
                    emit notificationDisplayed(widget->data.id, widget->data.app_name,
                                               widget->data.summary, widget->data.body);
                });
            }
            emit widget->displayed();
        } else {
            logger.debug(
                QString("No screen space for notification %1. Queuing").arg(widget->data.id));
        }
    }
}

void Notifier::closeNotification(int notification_id, NotificationCloseReason reason,
                                 bool is_batch) {
    auto it = m_notification_widgets.find({m_run_id, notification_id});
    if (it == m_notification_widgets.end()) {
        logger.error(QString("Could not find notification with id %1").arg(notification_id));
        return;
    }
    closeNotification(it->second, reason, is_batch);
}

void Notifier::closeNotification(NotificationWidget* widget, NotificationCloseReason reason,
                                 bool is_batch) {
    auto key = std::make_pair(widget->data.notification_tray_run_id, widget->data.id);

    // Check if this notification has already been closed and removed
    auto it = m_notification_widgets.find(key);
    if (it == m_notification_widgets.end()) {
        logger.debug(QString("Notification %1 already closed, skipping").arg(widget->data.id));
        return;
    }

    logger.info(QString("Closing notification %1").arg(widget->data.id));

    if (widget->isVisible()) {
        widget->close();
    }

    if (reason != NotificationCloseReason::CLOSED_BY_CALL_TO_CLOSENOTIFICATION) {
        emit notificationClosed(widget->data.id, static_cast<int>(reason),
                                QString::fromStdString(widget->data.path.string()),
                                widget->data.notification_tray_run_id == m_run_id);
    }

    // Remove the widget from the map and schedule it for deletion
    m_notification_widgets.erase(it);
    widget->deleteLater();

    // Recalculate offset based on remaining visible widgets
    m_offset = 0;
    for (auto& [k, w] : m_notification_widgets) {
        if (w->isVisible()) {
            m_offset += w->height() + 10;
        }
    }

    // Show queued notifications that weren't displayed yet
    for (auto& [k, w] : m_notification_widgets) {
        if (!w->was_displayed) {
            showOrQueueNotification(w);
        }
    }
}

void Notifier::batchNotify() {
    // Implementation for batch notification processing
    // Includes post-DnD catchup logic from Python version

    std::function<void(NotificationFolder&)> process = [&](NotificationFolder& folder) {
        std::vector<CachedNotification> new_notifications;

        // Initialize last_notified for this folder if not present
        if (m_last_notified.find(folder.path) == m_last_notified.end()) {
            m_last_notified[folder.path] = -1;
        }

        for (auto& [name, notif] : folder.notifications) {
            if (notif.trashed)
                continue;

            int backoff = Settings::getNotificationBackoffMinutes(m_root_path, folder.path,
                                                                  m_notification_backoff_minutes);
            qint64 minutes_since = notif.at.secsTo(QDateTime::currentDateTimeUtc()) / 60;

            bool dnd_active =
                Settings::isDoNotDisturbActive(m_root_path, folder.path, m_do_not_disturb);

            if (!dnd_active) {
                bool should_notify = false;

                // Check if within backoff window
                if (backoff > 0 && minutes_since <= backoff) {
                    should_notify = true;
                }

                // Check if we just came back from DnD (post-DnD catchup)
                auto do_not_disturb =
                    Settings::getDoNotDisturb(m_root_path, folder.path, m_do_not_disturb);
                if (do_not_disturb.has_value()) {
                    QDateTime dnd_end = do_not_disturb.value();
                    // DnD ended after we started AND notification arrived after DnD ended
                    // AND we haven't notified about this yet
                    if (dnd_end >= m_started_at && notif.at >= dnd_end &&
                        notif.id > m_last_notified[folder.path]) {
                        should_notify = true;
                    }
                }

                if (should_notify) {
                    new_notifications.push_back(notif);
                }
            }
        }

        if (!new_notifications.empty()) {
            notify(new_notifications, true);
        }

        for (auto& [name, subfolder] : folder.folders) {
            process(subfolder);
        }
    };

    process(m_notification_cache);
}

void Notifier::playNotificationSound(const CachedNotification& notification) {
    if (notification.hints.value("suppress-sound", false).toBool()) {
        logger.debug(QString("Notification %1 has suppress-sound hint").arg(notification.id));
        return;
    }

    QString audio_path;

    if (notification.hints.contains("sound-file")) {
        audio_path = notification.hints["sound-file"].toString();
    } else if (notification.hints.contains("sound-name")) {
        QString sound_name = notification.hints["sound-name"].toString();
        audio_path = QString("/usr/share/sounds/freedesktop/%1.oga").arg(sound_name);
    } else {
        fs::path current = notification.path.parent_path();
        while (current != m_root_path.parent_path()) {
            fs::path sound_file = current / ".notification.wav";
            if (fs::exists(sound_file)) {
                audio_path = QString::fromStdString(sound_file.string());
                break;
            }
            current = current.parent_path();
        }
    }

    if (!audio_path.isEmpty() && QFile::exists(audio_path)) {
        logger.debug(QString("Playing notification sound: %1").arg(audio_path));
        QSoundEffect* sound = new QSoundEffect(this);
        sound->setSource(QUrl::fromLocalFile(audio_path));
        sound->play();
        // Clean up after playing
        connect(sound, &QSoundEffect::playingChanged, [sound]() {
            if (!sound->isPlaying()) {
                sound->deleteLater();
            }
        });
    }
}

void Notifier::snoozeNotification(NotificationWidget* widget, int duration_ms) {
    logger.info(QString("Snoozing notification %1 for %2 seconds")
                    .arg(widget->data.id)
                    .arg(duration_ms / 1000.0));

    // Create a timer to re-display the notification
    QTimer* timer = new QTimer(this);
    timer->setSingleShot(true);

    // Copy the notification data
    CachedNotification notification_data = widget->data;

    // When timer expires, create a new notification widget and display it
    connect(timer, &QTimer::timeout, [this, notification_data]() {
        logger.info(QString("Re-displaying snoozed notification %1").arg(notification_data.id));

        NotificationWidget* new_widget = new NotificationWidget(notification_data);

        connect(new_widget, &NotificationWidget::closed, [this, new_widget](int reason) {
            closeNotification(new_widget, static_cast<NotificationCloseReason>(reason), false);
        });

        connect(new_widget, &NotificationWidget::snoozed, [this, new_widget](int duration_ms) {
            snoozeNotification(new_widget, duration_ms);
        });

        if (new_widget->data.notification_tray_run_id == m_run_id) {
            connect(new_widget, &NotificationWidget::actionInvoked,
                    [this, new_widget](const QString& key) {
                        emit actionInvoked(new_widget->data.id, key);
                    });
        }

        showOrQueueNotification(new_widget);
    });

    timer->start(duration_ms);
}
